core/trino-main/pom.xml

```diff
@@ -396,6 +396,25 @@
             <scope>test</scope>
         </dependency>
 
+        <dependency>
+            <groupId>org.junit.jupiter</groupId>
+            <artifactId>junit-jupiter-api</artifactId>
```

Comment:
After this change, trino-main will have two test libraries on the classpath, and so we should expect sooner or later we will have test classes accidentally mixing annotations coming from the two.

Code review process won't be able to identify such cases, and we may eg start leaking resources due to cleanup methods not getting called.

Before we do this here, it's paramount that we provide static or runtime validation detecting such problems.
See https://github.com/trinodb/trino/issues/9378 for checkers that needs to introduced.

Reply:
@findepi I see, and what would be the best way to introduce those checkers mentioned in the #9378 in the updated test classes ? should I use an if statement similar to this one below : 
```

if (a instanceof X) {
    // do something
  }
```

Reply:
@findepi, I don’t think that’s a real issue, so while having a checker is nice, it’s not necessarily a pre-requisite, especially if we migrate “bubbles” of related tests at a time.

We already have junit 4 in the classpath and it has never been a problem.

Reply:
> We should now find out whether JUnit allows similar verifications. (If it doesn't, we shouldn't move over to it, but that's stating the obvious).

That's not at all obvious. JUnit has some features that would allow us to write and structure tests in a better way (especially around how to inject and manage lifecycle of resources that are external to the tests) that would make it worth on their own, even if we don't have those checkers -- FWIW, the project survived for more than 7 years without such checkers without many issues in practice.

Regardless, JUnit does have mechanisms for creating listeners and intercepting test invocations.

Reply:
> TestWatcher interface is implemented differently as compared to those 2 interfaces available from testNg, replacing the existing checkers is likely going to involve writing new checkers from scratch.

understood and expected. the checkers' logic is thin as compared with technical aspects, and we should port them over
let's not waste time trying to unify the code, or anything like that

> Lemme know if this is something we wanna consider.

absolutely!
thanks for finding the way forward

Reply:
Sure, not a problem. 

However, the task of writing these new checkers using JUnit from scratch seems a bit daunting to me. But I will try wrapping my head around the existing checkers and see how far I can get with implementing them in JUnit. But, yea, I don't feel very confident about being able to do it all by myself. 

Would anyone be able to suggest an approach I can take to get started with re writing a checker, and probably point out some low hanging fruits in the Trino code base which I can target to gain confidence as I move forward ? 

Also, who among the maintainers would be the best person to ask questions related to JUnit and checkers ? I might have doubts when trying to write them.